File: src/tutorial_py_pkg/tutorial_py_pkg/service_client.py
import rclpy
from rclpy.node import Node
from functools import partial
from tutorial_cpp_pkg.srv import TestService

class MyNode(Node):
    def __init__(self):
        super().__init__("Python_Service_Client")
        self.get_logger().info("Start client...")
        # self.callService([2.3, 4.5])
        
    def callService(self, datas):
        client = self.create_client(TestService, "py_tutorial_service")
        while not client.wait_for_service(1.0):
            self.get_logger().warn("Waiting for Server ...")
            
        request = TestService.Request()
        request.datas = datas
        
        self.get_logger().info("send request")
        future = client.call_async(request)
        future.add_done_callback(partial(self.callback_callService))
        
    def callback_callService(self, future):
        try:
            response = future.result()
            self.get_logger().info("result: " + str(response.sum))
        except Exception as e:
            self.get_logger().error("Service call Failed %r" % (e,))
            

def main(args=None):
    rclpy.init(args=args)
    node = MyNode()
    rclpy.spin(node)
    rclpy.shutdown()
    
if __name__ == "__main__":
    main()

File: src/tutorial_cpp_pkg/src/service_client.cpp
#include "rclcpp/rclcpp.hpp"
#include "tutorial_cpp_pkg/srv/test_service.hpp"

class MyNode: public rclcpp::Node
{
public:
    MyNode() : Node("CPP_Service_Client") {
        RCLCPP_INFO(this->get_logger(), "Cpp Service Client");
        this->declare_parameter("param1", 1.1);
        this->declare_parameter("param2", 2.2);
        thread_ = std::thread(std::bind(&MyNode::callService, this, std::vector<double>{this->get_parameter("param1").as_double(), this->get_parameter("param2").as_double()}));
        
        RCLCPP_INFO(this->get_logger(), "Cpp Service Client inited");
    }

private:
    void callService(std::vector<double> datas) {
        RCLCPP_INFO(this->get_logger(), "callService");
        auto client = this->create_client<tutorial_cpp_pkg::srv::TestService>("cpp_tutorial_service");
        while(!client->wait_for_service(std::chrono::seconds(1))) {
            RCLCPP_WARN(this->get_logger(), "Waiting for Server...");
        }

        auto request = std::make_shared<tutorial_cpp_pkg::srv::TestService::Request>();
        request->datas = datas;

        auto future = client->async_send_request(request);

        try {
            auto response = future.get();
            RCLCPP_INFO(this->get_logger(), "Result: %lf", response->sum);
        }
        catch (const std::exception &e) {
            RCLCPP_ERROR(this->get_logger(), "Service call failed");
        }
    }

private:
    std::thread thread_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<MyNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}